### PR TITLE
[SPARK-36100][CORE] Grouping exception in core/status

### PR DESCRIPTION
**Pull request by dgd-contributor**
### What changes were proposed in this pull request?
This PR group exception messages in core/src/main/scala/org/apache/spark/status

### Why are the changes needed?
It will largely help with standardization of error messages and its maintenance.

### Does this PR introduce _any_ user-facing change?
No. Error messages remain unchanged.

### How was this patch tested?
No new tests - pass all original tests to make sure it doesn't break any existing behavior.

### CODE REVIEW

**Comment by mridulm** [comment]
I have not followed on this work, but looking at this pr a general comment, please use as specific a subclass as possible without loosing generalizability while defining interfaces.
For example, use Exception instead of Throwable - or relevant subclass is possible.

**Comment by cloud-fan** [comment]
> please use as specific a subclass as possible without loosing generalizability while defining interfaces.

Yea I think this is the policy. The recent work is just to centralize all the errors into one or a few files. The error message and exception class should not be changed for now.

File: core/src/main/scala/org/apache/spark/errors/SparkCoreErrors.scala [hunk 1]
```diff
@@ -141,4 +146,102 @@ object SparkCoreErrors {
...
+
+  def indexOutOfBoundError(idx: Int): Throwable = {
+    new IndexOutOfBoundsException(idx.toString)
+  }
+
+  def notAuthorizedUserError(user: String): Throwable = {
+    new ForbiddenException(s"""user "$user" is not authorized""")
+  }
+
+  def notFoundAppKeyError(appKey: String): Throwable = {
```
Comment (allisonwang-db) [inline core/src/main/scala/org/apache/spark/errors/SparkCoreErrors.scala:176]: appNotFoundError

File: core/src/main/scala/org/apache/spark/errors/SparkCoreErrors.scala [hunk 2]
```diff
@@ -141,4 +146,102 @@ object SparkCoreErrors {
...
+    new NotFoundException(s"no such app: $appKey")
+  }
+
+  def notFoundJobIdError(jobId: Int): Throwable = {
```
Comment (allisonwang-db) [inline core/src/main/scala/org/apache/spark/errors/SparkCoreErrors.scala:180]: unknownJobError

File: core/src/main/scala/org/apache/spark/errors/SparkCoreErrors.scala [hunk 3]
```diff
@@ -141,4 +146,102 @@ object SparkCoreErrors {
...
+
+  def threadDumpsNotAvailableError(): Throwable = {
+    new ServiceUnavailable("Thread dumps not available through the history server.")
+  }
+
+  def noThreadDumpAvailableError(): Throwable = {
+    new NotFoundException("No thread dump is available.")
+  }
+
+  def notFoundHttpRequestError(uri: String): Throwable = {
```
Comment (allisonwang-db) [inline core/src/main/scala/org/apache/spark/errors/SparkCoreErrors.scala:196]: uriNotFoundError

File: core/src/main/scala/org/apache/spark/errors/SparkCoreErrors.scala [hunk 4]
```diff
@@ -141,4 +146,102 @@ object SparkCoreErrors {
...
+
+  def eventLogsNotAvailableError(appId: String): Throwable = {
+    new ServiceUnavailable(s"Event logs are not available for app: $appId.")
+  }
+
+  def unknownAppError(appId: String): Throwable = {
+    new NotFoundException(s"unknown app: $appId")
+  }
+
+  def notFoundAppWithAttemptError(appId: String, attemptId: String): Throwable = {
```
Comment (allisonwang-db) [inline core/src/main/scala/org/apache/spark/errors/SparkCoreErrors.scala:220]: unknownAppWithAttemptError

File: core/src/main/scala/org/apache/spark/errors/SparkCoreErrors.scala [hunk 5]
```diff
@@ -141,4 +146,102 @@ object SparkCoreErrors {
...
+
+  def unknownStageError(stageId: Int): Throwable = {
+    new NotFoundException(s"unknown stage: $stageId")
+  }
+
+  def unknownAttemptForStageError(stageId: Int, msg: String): Throwable = {
+    new NotFoundException(s"unknown attempt for stage $stageId.  Found attempts: [$msg]")
+  }
+
+  def noTasksReportMetricsError(stageId: Int, stageAttemptId: Int): Throwable = {
```
Comment (allisonwang-db) [inline core/src/main/scala/org/apache/spark/errors/SparkCoreErrors.scala:232]: noTaskReportedMetricsError

File: core/src/main/scala/org/apache/spark/errors/SparkCoreErrors.scala [hunk 6]
```diff
@@ -141,4 +146,102 @@ object SparkCoreErrors {
...
+    new NotFoundException(s"No tasks reported metrics for $stageId / $stageAttemptId yet.")
+  }
+
+  def badParameterErrors(param: String, exp: String, actual: String): Throwable = {
```
Comment (allisonwang-db) [inline core/src/main/scala/org/apache/spark/errors/SparkCoreErrors.scala:236]: badParameterError

**Review by allisonwang-db — APPROVED** [review]
Looks good!

**Comment by dgd-contributor** [comment]
Thanks for your reviews.

**Comment by dgd-contributor** [comment]
Please see PR [#34191](https://github.com/apache/spark/pull/34191)

**Pull request closed without merging**